### PR TITLE
DAOS-15044: Refactor error codes

### PR DESCRIPTION
**Pull request by frostedcmos**
- Most mercury error codes will not map to DER_HG, resulting in retry
- crt_hg_bulk_create() and access() modified to not return HG error

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/cart/crt_hg.h [hunk 1]
```diff
@@ -193,11 +194,11 @@ crt_hgret_2_der(int hg_ret)
 	case HG_BUSY:
 		return -DER_BUSY;
 	case HG_FAULT:
-	case HG_PROTOCOL_ERROR:
 		return -DER_HG_FATAL;
 	case HG_PERMISSION:
 	case HG_ACCESS:
 		return -DER_NO_PERM;
+	case HG_PROTOCOL_ERROR:
```
Comment (liw) [inline src/cart/crt_hg.h:201]: While I'm okay with this, I think a better solution to assess in what cases Mercury should not return this error. See [my comment today on the Jira ticket](https://daosio.atlassian.net/browse/DAOS-15044?focusedCommentId=124785).

**Review by liw — APPROVED** [review]

File: src/cart/crt_hg.h (hunk 1 above)
Comment (liw) [reply src/cart/crt_hg.h:201]: Or, perhaps these two approaches are both necessary...

Comment (frostedcmos) [reply src/cart/crt_hg.h:201]: As per discussion with Jerome, this error is often returned by mercury as a result of 'unknown error' type in the lower levels (fabric and ucx), such as something failed but didnt specify the reason. As such this error might be retriable in some contexts (such as send failure) but not in others (such as bulk allocation failure).

**Review by soumagne — APPROVED** [review]
it's fine to merge it but we'll have to revisit that imo and group some type of errors as fatal errors.

**Pull request merged**